### PR TITLE
Dedup more environment code

### PR DESCRIPTION
**Pull request by pattyshack**
NOTE:
 1. I've move meterMemory's logic into MeterMemory.
 2. GetStorageCapacity: we're using the script env version, which has additional
    comments.
 3. GetAccountAvailableBalance: we're using the script env verion, which returns
    0 instead of "value" (which is also zero) on error.
 4. GetAccountKey: we're using the transaction env version.  accountKeys is
    never nil.

### CODE REVIEW

File: fvm/env.go [hunk 1]
```diff
@@ -283,6 +284,75 @@ func storageMBUFixToBytesUInt(result cadence.Value) uint64 {
...
+	address common.Address,
+) (
+	value uint64,
+	err error,
+) {
+	defer env.StartSpanFromRoot(trace.FVMEnvGetAccountBalance).End()
+
+	err = env.Meter(meter.ComputationKindGetAccountBalance, 1)
+	if err != nil {
+		return value, fmt.Errorf("get account balance failed: %w", err)
```
Comment (SaveTheRbtz) [inline fvm/env.go:323]:
```suggestion
		return 0, fmt.Errorf("get account balance failed: %w", err)
```

**Review by SaveTheRbtz — APPROVED** [review]

**Review by janezpodhostnik — APPROVED** [review]

File: fvm/env.go [hunk 2]
```diff
@@ -283,6 +284,75 @@ func storageMBUFixToBytesUInt(result cadence.Value) uint64 {
...
+	address common.Address,
+) (
+	value uint64,
+	err error,
+) {
+	defer env.StartSpanFromRoot(trace.FVMEnvGetStorageCapacity).End()
+
+	err = env.Meter(meter.ComputationKindGetStorageCapacity, 1)
+	if err != nil {
+		return 0, fmt.Errorf("get storage capacity failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:297]: Does `Meter` actually get the storage capacity or just increment the meter by 1?

If is later, it's better to mention that it's the metering failed not the actual data query failed

```suggestion
		return 0, fmt.Errorf("meter get storage capacity failed: %w", err)
```

File: fvm/env.go (hunk 1 above)
Comment (zhangchiqing) [inline fvm/env.go:323]:
```suggestion
		return value, fmt.Errorf("meter get account balance failed: %w", err)
```

File: fvm/env.go [hunk 3]
```diff
@@ -301,6 +371,80 @@ func (env *commonEnv) GetAccountContractNames(address runtime.Address) ([]string
...
 
+func (env *commonEnv) ResolveLocation(
+	identifiers []runtime.Identifier,
+	location runtime.Location,
+) ([]runtime.ResolvedLocation, error) {
+	defer env.StartExtensiveTracingSpanFromRoot(trace.FVMEnvResolveLocation).End()
+
+	err := env.Meter(meter.ComputationKindResolveLocation, 1)
+	if err != nil {
+		return nil, fmt.Errorf("resolve location failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:382]:
```suggestion
		return nil, fmt.Errorf("meter resolve location failed: %w", err)
```

File: fvm/env.go [hunk 4]
```diff
@@ -301,6 +371,80 @@ func (env *commonEnv) GetAccountContractNames(address runtime.Address) ([]string
...
+
+	// if the location is an address,
+	// and no specific identifiers where requested in the import statement,
+	// then fetch all identifiers at this address
+	if len(identifiers) == 0 {
+		address := flow.Address(addressLocation.Address)
+
+		err := env.accounts.CheckAccountNotFrozen(address)
+		if err != nil {
+			return nil, fmt.Errorf("resolving location failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:407]:
```suggestion
			return nil, fmt.Errorf("checking account not frozen failed: %w", err)
```

File: fvm/env.go [hunk 5]
```diff
@@ -436,6 +580,49 @@ func (env *commonEnv) Commit() (programs.ModifiedSets, error) {
...
+	hashAlgorithm runtime.HashAlgorithm,
+) (
+	bool,
+	error,
+) {
+	defer env.StartSpanFromRoot(trace.FVMEnvVerifySignature).End()
+
+	err := env.Meter(meter.ComputationKindVerifySignature, 1)
+	if err != nil {
+		return false, fmt.Errorf("verify signature failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:598]:
```suggestion
		return false, fmt.Errorf("meter verify signature failed: %w", err)
```

File: fvm/env.go [hunk 6]
```diff
@@ -436,6 +580,49 @@ func (env *commonEnv) Commit() (programs.ModifiedSets, error) {
...
+	}
+
+	return valid, nil
+}
+
+func (env *commonEnv) ValidatePublicKey(pk *runtime.PublicKey) error {
+	err := env.Meter(meter.ComputationKindValidatePublicKey, 1)
+	if err != nil {
+		return fmt.Errorf("validate public key failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:620]:
```suggestion
		return fmt.Errorf("meter validate public key failed: %w", err)
```

File: fvm/env.go [hunk 7]
```diff
@@ -472,3 +659,29 @@ func (env *commonEnv) AllocateStorageIndex(owner []byte) (atree.StorageIndex, er
...
+	keyIndex int,
+) (
+	*runtime.AccountKey,
+	error,
+) {
+	defer env.StartSpanFromRoot(trace.FVMEnvGetAccountKey).End()
+
+	err := env.Meter(meter.ComputationKindGetAccountKey, 1)
+	if err != nil {
+		return nil, fmt.Errorf("get account key failed: %w", err)
```
Comment (zhangchiqing) [inline fvm/env.go:679]:
```suggestion
		return nil, fmt.Errorf("meter get account key failed: %w", err)
```

**Review by zhangchiqing — APPROVED** [review]
Nice refactor. 

I just added a few comments for better error messages.

File: fvm/env.go (hunk 2 above)
Comment (pattyshack) [reply fvm/env.go:298]: the error's detail is specify at the end.  for metering error, the specific string is "computation limit exceed ...".  It should be sufficiently clear.

File: fvm/env.go (hunk 4 above)
Comment (pattyshack) [reply fvm/env.go:408]: updated this error string since it's less clear.

**Comment by pattyshack** [comment]
bors merge

**Pull request merged**